### PR TITLE
Add Detox for e2e testing

### PR DESCRIPTION
**Pull request by hawkrives**
Detox is an end-to-end testing platform thingy.

It can do a lot ([actions](https://wix.github.io/detox/docs/APIRef.ActionsOnElement.html#swipedirection-speed-percentage), for instance, like swiping, or tapping, or long-pressing, or shaking; or [performance testing](https://github.com/wix/DetoxInstruments)). There's also an [overview page](https://github.com/wix/detox/blob/master/docs/Introduction.HowDetoxWorks.md) talking about how it works, technically.

---

Things this doesn't do, but could in the future:

- run on CI
- take screenshots
- take videos(!)
- open test-specific screens (think testing the scrolling-textinputs behavior)
- run on Android (should be fairly easy)
- contribute to code coverage (??? idk)
- performance testing / benchmarking (https://github.com/wix/DetoxInstruments)

Things this does do:

- assert that the app launches and shows the homescreen on iOS
- assert that the app can open the Settings screen
- assert that the app can navigate from Home > Settings > Home

---

To run detox, run `yarn detox build`, to build your app and install it, then run `yarn detox test` to actually execute the tests. You won't need to re-run `build` until you change the native code (same as normal, interactive testing).

I think there's a way to tell it what platform to test, but I haven't dug in that far yet.

The test uses the iPhone 7 simulator; I don't know what iOS revision it wants, but mine picked iOS 11.2. If you get odd errors, please say so?

---

The other part of this PR is fixing up a few voiceover/accessibility things; namely giving proper labels to a few buttons. For instance, the "settings" button will now be read as "Open Settings, button" instead of … trying to read a Unicode private-use character. I'm not sure how it would have read that, but this has _got_ to be nicer.

I think that we'll get much better about accessibility by adding e2e tests, judging from my experience today.

---

This is not part of any milestone yet, because I kinda did this on a whim, and I don't know what the reaction will be, so I don't have a target merge window yet.

### CODE REVIEW

File: e2e/config.json [hunk 1]
```diff
@@ -0,0 +1,4 @@
+{
+    "setupTestFrameworkScriptFile": "./init.js",
```
Comment (rye) [inline e2e/config.json]: Please use two-space indentation instead of four-space per the general rule of thumb with JSON in this repo.

File: source/app.js [hunk 2]
```diff
@@ -30,6 +30,8 @@ function getCurrentRouteName(navigationState: NavigationState): ?string {
 
 type Props = {}
 
+const navigationPersistenceKey = __DEV__ ? null : 'NavState'
```
Comment (rye) [inline source/app.js:33]: This seems out of scope for this PR to me—this fix could get pulled out into its own PR.

**Review by rye — CHANGES_REQUESTED** [review]
All in all, I'm okay with this. I really would just like the four-space indentation resolved. `navigationPersistenceKey` extraction seems out of scope to me, but if that was bundled into this PR I wouldn't not be happy.

**Comment by rye** [comment]
Why `iPhone 7`, specifically? How much is required for Android to join along? I could easily see scaling these tests out.

Also, is there any way to strip out `testID`s from the resulting build? Or are those benign?

File: e2e/config.json (hunk 1 above)
Comment (hawkrives) [reply e2e/config.json]: Fixed. (by using tabs, which are only not used in package.json because npm/yarn reformat it every time they run.)

**Comment by hawkrives** [comment]
> Why iPhone 7, specifically? 

It was in their example, and for the initial POC pass I didn't see an advantage to running on several devices. We can add new config keys to run on other devices, and it can run them in parallel.

> How much is required for Android to join along? I could easily see scaling these tests out.

https://wix.github.io/detox/docs/Introduction.Android.html

tl;dr: pretty much add a `test:app` gradle target, add a new detox config key, and run `yarn detox test -c android.emu.debug`.

> Also, is there any way to strip out testIDs from the resulting build? Or are those benign?

Completely benign. They're pretty much just like `id` attributes in HTML, as far as the app goes. I don't even know if they get propagated to the native layers.

File: source/app.js (hunk 2 above)
Comment (hawkrives) [reply source/app.js:33]: Done (#2813).

I'm leaving it in here until it's merged, so that the tests in this PR actually pass.

File: e2e/config.json [hunk 3]
```diff
@@ -0,0 +1,4 @@
+{
+	"setupTestFrameworkScriptFile": "./init.js",
```
Comment (rye) [inline e2e/config.json:2]: Do we want tabs? Or two-space as is the custom with `package.json`?

Comment (hawkrives) [reply e2e/config.json:2]: Tabs for consistency with everything that's not package.json

Comment (hawkrives) [reply e2e/config.json:2]: (Plus, tab is 1 char, 2 spaces is 2 chars. 1 < 2, so 1 is better :p)

Comment (rye) [reply e2e/config.json:2]: Alright, I will not argue with something that I agree with. Tabs are better.

**Review by rye — DISMISSED** [review]

**Comment by hawkrives** [comment]
@drewvolz any opposition? I know we'll add more stuff later, but I don't know that we need any additions before we introduce this to our codebase.

File: e2e/firstTest.spec.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+/* eslint-env jest */
+/* global element, device, by */
+
+describe('Example', () => {
```
Comment (drewvolz) [inline e2e/firstTest.spec.js]: Please rename the describe to something other than `Example`

File: source/app.js (hunk 2 above)
Comment (drewvolz) [reply source/app.js:33]: That has been merged so this may be removed.

**Review by rye — APPROVED** [review]

**Review by drewvolz — APPROVED** [review]
